### PR TITLE
feat(pipeline): multi-branch pipelines

### LINKED ISSUE #302: [FEATURE] Pipeline per branch
### Describe the feature

Allow PDKPipelineProjects to create and destroy pipelines as git branches are created and destroyed. Alternatively, enable additional pipelines tied to a branch per developer.

### Use Case

When developing as a team, we often work on different features in branches. We need to test deployments before merging to the main branch, so we currently deploy manually to our individual accounts. This will allow us to test by pushing a commit without impacting the main branch or our teammates' work.

### Proposed Solution

Amplify's CI/CD monitors the creation and deletion of branches to create new environments with their own pipelines. PDKPipeline could create and eventbridge rule to trigger pipeline stack creation and destruction when codecommit branches are created and destroyed.

### Other Information

Blocked by #239

### Acknowledgements

- [x] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PDK version used

0.14.8

### What languages will this feature affect?

_No response_

### Environment details (OS name and version, etc.)

MacOS

### PR DESCRIPTION
**Pull request by jstrunk**
If your team follows [GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow), the pipelines module can optionally help you create independent environments to test and validate changes before merging. When you create a new branch, it will automatically create a new pipeline stack and any stages you configure. Once you have finished testing and delete the branch, it will automatically clean up the stacks created in the branch's environment.

This feature is based on the architecture in [Multi-branch pipeline management and infrastructure deployment using AWS CDK Pipelines](https://aws.amazon.com/blogs/devops/multi-branch-pipeline-management-and-infrastructure-deployment-using-aws-cdk-pipelines/) . EventBridge Rules trigger Lambda functions on CodeCommit repo branch creation and deletion. The Lambda function for creation sets the `BRANCH` environment variable and runs a CodeBuild project that creates the pipeline stack for the branch. It is up to the developer to give their pipeline stack and stages unique names per branch. The examples show how to use the `BRANCH` variable to do so. The multi-branch stacks are tagged with the branch name and repository name. The Lambda for branch deletion lists all stacks matching those tags and deletes them.

I tested this change using a Typescript pipeline project.

Resolves #302

### CODE REVIEW

File: packages/pipeline/README.md
Comment (cogwirrel) [inline packages/pipeline/README.md]: Nit: *your pipeline :)

File: packages/pipeline/test/pdk-pipeline.test.ts
Comment (cogwirrel) [inline packages/pipeline/test/pdk-pipeline.test.ts]: It might be worth cleaning up the environment in a `finally` block to reduce the chances of a failure in one test affecting another :)

File: packages/pipeline/README.md
Comment (cogwirrel) [inline packages/pipeline/README.md]: It might be a little more readable if we exposed a constant like `PDKPipeline.ALL_BRANCHES`? `[""]` makes logical sense but is a little harder to reason about at first glance :)

File: packages/pipeline/src/utils.ts
Comment (cogwirrel) [inline packages/pipeline/src/utils.ts]: Nit: it might be neater to have these as static members of the `PDKPipeline` construct rather than separate them out as they're quite closely related :)

File: packages/pipeline/src/feature-branches.ts
Comment (cogwirrel) [inline packages/pipeline/src/feature-branches.ts]: The user can pass custom install/build commands to `PDKPipeline`, and so I think we want to honour those here too?

File: packages/pipeline/src/feature-branches.ts
Comment (cogwirrel) [inline packages/pipeline/src/feature-branches.ts]: Nit: looks like this prop is required so doesn't have a default here :)

File: packages/pipeline/samples/typescript/src/pipeline.ts
Comment (cogwirrel) [inline packages/pipeline/samples/typescript/src/pipeline.ts]: It might be nicer to abstract away the environment variable from users, and provide some helpers like `PDKPipeline.getBranchPrefix(app)` and `PDKPipeline.isDefaultBranch(app)`?

File: packages/pipeline/src/feature-branches.ts
Comment (cogwirrel) [inline packages/pipeline/src/feature-branches.ts]: Nit: `.map((prefix) => ({ prefix }))`

**Review by cogwirrel — COMMENTED** [review]
Very cool feature, nice job! 🚀

**Comment by JeremyJonas** [comment]
@jstrunk What is the status of this PR?

**Comment by jstrunk** [comment]
I am going to implement the comments, but I didn't have time last week. I
expect to make time for it this week.

File: packages/pipeline/src/feature-branches.ts
Comment (jstrunk) [reply packages/pipeline/src/feature-branches.ts]: I think it will work to replace only `"npx nx run-many --target=build --all"` with `commands` from `PDKPipeline`. The user wouldn't pass installation commands to `commands` in `synthShellStepPartialProps`.

File: packages/pipeline/test/pdk-pipeline.test.ts
Comment (jstrunk) [reply packages/pipeline/test/pdk-pipeline.test.ts]: I implemented an `afterEach` function to restore `process.env`.

**Comment by jstrunk** [comment]
I've addressed the comments. Tomorrow I will update the samples.

File: packages/pipeline/src/feature-branches.ts
Comment (cogwirrel) [inline packages/pipeline/src/feature-branches.ts]: We should allow the user to override the install commands too as not everyone will use `yarn`, ie some may use `pnpm` or `npm` :) They're currently overideable in `PDKPipeline` via `synthShellStepPartialProps` so I think we can pass whatever value that has if present, eg `synthShellStepPartialProps?.installCommands ?? ['npm install -g aws-cdk', 'yarn ....']`

File: packages/pipeline/src/feature-branches.ts
Comment (cogwirrel) [inline packages/pipeline/src/feature-branches.ts]: Technically depending on the package manager one might not always use `npx` (eg if you're using pnpm the command might be `pnpm dlx`). I wonder if we should also allow the user to override default deploy commands?

File: packages/pipeline/src/pdk-pipeline.ts
Comment (cogwirrel) [inline packages/pipeline/src/pdk-pipeline.ts:166]: Nice! 😄

**Review by cogwirrel — COMMENTED** [review]
Looks great, couple of minor comments :)

File: packages/pipeline/src/feature-branches.ts
Comment (cogwirrel) [reply packages/pipeline/src/feature-branches.ts]: You can pass `installCommands` to `synthShellStepPartialProps`, which is particularly useful if you're using a different package manager to yarn, like pnpm :) So I think we definitely want to honour what's configured in `PDKPipeline` where possible.

Comment (cogwirrel) [reply packages/pipeline/src/feature-branches.ts]: You can pass `installCommands` to `synthShellStepPartialProps`, which is particularly useful if you're using a different package manager to yarn, like pnpm :) So I think we definitely want to honour what's configured in `PDKPipeline` where possible.

File: packages/pipeline/src/feature-branches.ts
Comment (jstrunk) [reply packages/pipeline/src/feature-branches.ts]: I added a `cdkCommand` prop to customize this.

File: packages/pipeline/src/feature-branches.ts
Comment (jstrunk) [reply packages/pipeline/src/feature-branches.ts]: I modified the construct to take `synthShellStepPartialProps` directly and get build command and install commands there.

**Comment by jstrunk** [comment]
I've implemented the latest comments.

**Review by cogwirrel — APPROVED** [review]
Great stuff! Thanks!

**Comment by agdimech** [comment]
@jstrunk - Hi Jeff. I am noticing a lot of conflicts and files that have changed that are not directly apart of this PR (due to the large amount of structural PRs that have been merged recently and how the rebase has occurred).

Are you able to just cherry pick your specific changes into a fresh branch? Otherwise I am more than happy to do this on your behalf if you prefer?

Apart from that all your code looks good and can be merged once we get the commit/merges cleaned up :)

**Comment by jstrunk** [comment]
> @jstrunk - Hi Jeff. I am noticing a lot of conflicts and files that have changed that are not directly apart of this PR (due to the large amount of structural PRs that have been merged recently and how the rebase has occurred).
> 
> Are you able to just cherry pick your specific changes into a fresh branch? Otherwise I am more than happy to do this on your behalf if you prefer?
> 
> Apart from that all your code looks good and can be merged once we get the commit/merges cleaned up :)

That makes sense. I don't think that should be to much trouble. The files that belong in the PR can be copied over whole.

If I do that, how should I associate that new branch with this pr?

**Comment by agdimech** [comment]
> > @jstrunk - Hi Jeff. I am noticing a lot of conflicts and files that have changed that are not directly apart of this PR (due to the large amount of structural PRs that have been merged recently and how the rebase has occurred).
> > Are you able to just cherry pick your specific changes into a fresh branch? Otherwise I am more than happy to do this on your behalf if you prefer?
> > Apart from that all your code looks good and can be merged once we get the commit/merges cleaned up :)
> 
> That makes sense. I don't think that should be to much trouble. The files that belong in the PR can be copied over whole.
> 
> If I do that, how should I associate that new branch with this pr?

You can just open a new PR, abandon this one and just add a reference to this PR in the description of the new PR for traceability :)

**Comment by jstrunk** [comment]
Please continue this discussion in #403 .

**Pull request closed without merging**
